packages/runtime/src/components/_internal/ComponentErrorBoundary.tsx

```diff
@@ -0,0 +1,68 @@
+import React from 'react';
+
+type Props = {
+  componentId: string;
+  onRef: (ele: HTMLElement) => void;
+  onError?: (error: Error | null) => void;
+  onRecoverFromError: () => void;
+};
+
+export default class ComponentErrorBoundary extends React.Component<
+  Props,
+  { error: Error | null; rerenderFlag: number }
+> {
+  ref = React.createRef<HTMLDivElement>();
+  constructor(props: Props) {
+    super(props);
+    this.state = { error: null, rerenderFlag: 1 };
+  }
+
+  static getDerivedStateFromError(error: Error) {
+    return { error };
+  }
+
+  componentDidMount() {
+    if (this.ref.current) {
+      this.props.onRef(this.ref.current);
+    }
+    this.props.onError?.(null);
+  }
+
+  componentDidCatch(error: Error) {
+    this.props.onError?.(error);
+  }
+
+  componentDidUpdate() {
+    if (this.ref.current) {
+      this.props.onRef(this.ref.current);
+    }
+  }
+
+  onRerender = () => {
+    this.setState({ error: null });
+    // update the ref after rerender
+    setTimeout(() => {
```

Comment:
is the second parameter(callback) of setState a good place to put this logic?

Reply:
ok